### PR TITLE
Add cloudflare_access_identity_provider datasource

### LINKED ISSUE #1225: Add `cloudflare_access_identity_provider` data source
### Current Terraform and Cloudflare provider version

```
% terraform -v 
Terraform v0.13.7
+ provider registry.terraform.io/cloudflare/cloudflare v3.1.0
+ provider registry.terraform.io/hashicorp/aws v3.60.0
+ provider registry.terraform.io/hashicorp/random v3.1.0
```

### Description

I request adding a data source named `cloudflare_access_identity_provider`, allowing one to lookup metadata of a previously configured identity provider via the friendly name.

### Use cases

Resources like `cloudflare_access_application` require the unique ID number of an identity provider (the `allowed_idps` input variable). In situations where the Identity Provider's unique ID number cannot be shared via the terraform remote state, it would be great to lookup this ID number via a data source.

The Cloudflare API supports lookup of this ID as shown here. 

https://api.cloudflare.com/#access-identity-providers-list-access-identity-providers

```
curl -X GET "https://api.cloudflare.com/client/v4/accounts/699d98642c564d2e855e9661899b7252/access/identity_providers" \
    -H "Authorization: Bearer XXXXXXXXXXX" \
    -H "Content-Type: application/json"
```

### Potential Terraform configuration

```hcl

resource "cloudflare_access_application" "main" {
  zone_id                   = "example.com"
  name                      = "name"
  domain                    = "name.example.com"
  type                      = "self_hosted"
  session_duration          = "24h"
  allowed_idps              = [data.cloudflare_access_identity_provider.main.id]
  auto_redirect_to_identity = true
}

data "cloudflare_access_identity_provider" "main" {
  name = "my-idp"
}
```

### References

_No response_

### PR DESCRIPTION
**Pull request by cucxabong**
Fix #1225 
This is my very first MR, if anything need to be done, please advising. Thank you

### CODE REVIEW

File: cloudflare/data_source_access_identity_provider.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+				ExactlyOneOf: []string{"zone_id", "account_id"},
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
```
Comment (jacobbednarz) [inline cloudflare/data_source_access_identity_provider.go:33]:
```suggestion
```

`id` isn't an exportable field as it's an internal identifer by Terraform. `id` is added to all resources for free.

File: cloudflare/data_source_access_identity_provider.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	var accessIdentityProvider cloudflare.AccessIdentityProvider
+	for _, provider := range providers {
+		if provider.Name == name {
+			accessIdentityProvider = provider
+			break
+		}
+	}
+
+	if accessIdentityProvider.ID == "" {
+		return fmt.Errorf("no Access Identity Provider matching name '%s'", name)
```
Comment (jacobbednarz) [inline cloudflare/data_source_access_identity_provider.go:73]: `%q` quotes the string value for you

```suggestion
		return fmt.Errorf("no Access Identity Provider matching name %q", name)
```

File: cloudflare/data_source_access_identity_provider_test.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	"fmt"
+	"os"
+	"regexp"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+func TestAccCloudflareAccessIdentityProviderDataSource_PreventZoneIdAndAccountIdConflicts(t *testing.T) {
+
```
Comment (jacobbednarz) [inline cloudflare/data_source_access_identity_provider_test.go:13]:
```suggestion
```

File: website/cloudflare.erb [hunk 4]
```diff
@@ -49,6 +49,9 @@
             <li<%= sidebar_current("docs-cloudflare-datasource-zones") %>>
                 <a href="/docs/providers/cloudflare/d/zones.html">cloudflare_zones</a>
             </li>
+            <li<%= sidebar_current("docs-cloudflare-datasource-access-identity-provider") %>>
```
Comment (jacobbednarz) [inline website/cloudflare.erb:52]: can we keep this alphabetised please?

File: cloudflare/provider.go [hunk 5]
```diff
@@ -116,6 +116,7 @@ func Provider() *schema.Provider {
 			"cloudflare_zones":                       dataSourceCloudflareZones(),
 			"cloudflare_zone":                        dataSourceCloudflareZone(),
 			"cloudflare_zone_dnssec":                 dataSourceCloudflareZoneDNSSEC(),
+			"cloudflare_access_identity_provider":    dataSourceCloudflareAccessIdentityProvider(),
```
Comment (jacobbednarz) [inline cloudflare/provider.go:119]: same as https://github.com/cloudflare/terraform-provider-cloudflare/pull/1300/files#r743364820, we'll sort this alphabetically please.

File: website/docs/d/access_identity_provider.html.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+layout: "cloudflare"
+page_title: "Cloudflare: cloudflare_access_identity_provider"
+sidebar_current: "docs-cloudflare-datasource-access-identity-provider"
+description: |-
+  Get information on a Cloudflare Access Identity Provider by name.
+---
+
+# cloudflare_access_identity_provider
+
+Use this data source to lookup an [Access Identity Providers][access_identity_provider_guide] by name
```
Comment (jacobbednarz) [inline website/docs/d/access_identity_provider.html.md:11]:
```suggestion
Use this data source to lookup an [Access Identity Provider][access_identity_provider_guide] by name
```

File: website/docs/d/access_identity_provider.html.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+}
+```
+
+## Argument Reference
+The following arguments are supported:
+
+-> **Note:** It's required that an `account_id` or `zone_id` is provided.
+
+* `account_id` - (Optional) The account for which to look for an Access Identity Provider (Conflicts with `zone_id`)
+
```
Comment (jacobbednarz) [inline website/docs/d/access_identity_provider.html.md:38]:
```suggestion
```

**Comment by jacobbednarz** [comment]
i've pushed up some nitpicky fixes so the last thing outstanding is adding a CHANGELOG entry (see notes at [changelog-process.md](https://github.com/cloudflare/terraform-provider-cloudflare/blob/master/docs/changelog-process.md)) and we can get this merged

this is a solid PR, thanks for it! 🥇

**Review by jacobbednarz — COMMENTED** [review]
i've pushed up some nitpicky fixes so the last thing outstanding is adding a CHANGELOG entry (see notes at [changelog-process.md](https://github.com/cloudflare/terraform-provider-cloudflare/blob/master/docs/changelog-process.md)) and we can get this merged

this is a solid PR, thanks for it! 🥇

**Comment by cucxabong** [comment]
Thank you very much for your feedback. I have added change log for this MR, please help to take a look

**Comment by quanwf** [comment]
Up to avoid stale.

**Comment by cucxabong** [comment]
up to avoid stale

**Comment by jacobbednarz** [comment]
sorry for the delay, would you mind updating this PR to match the conventions in #1321 (with the schema in a dedicated file). after that, we're good to merge here.

**Comment by cucxabong** [comment]
updated, please take a look

**Comment by jacobbednarz** [comment]
thank you for your patience! 🎂

**Comment by cucxabong** [comment]
thanks for your help & congrats to you & your family 🍼 👶

**Pull request merged**